src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    /// <summary>
+    /// CA1849: Avoid constant arrays as arguments. Replace with static readonly arrays.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class AvoidConstArraysAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1849";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidConstArraysTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidConstArraysMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidConstArraysDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        private const string s_readonlySpanTypeName = WellKnownTypeNames.SystemReadOnlySpan1;
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            // Analyzes an argument operation
+            context.RegisterOperationAction(operationContext =>
+            {
+                IArgumentOperation? argumentOperation;
+                string readonlySpanTypeString = WellKnownTypeProvider.GetOrCreate(operationContext.Compilation)
+                    .GetOrCreateTypeByMetadataName(s_readonlySpanTypeName)!.Name;
```

Comment:
Better to keep it as a symbol instead of getting the name, and do a symbol equality instead of string equality.

Reply:
https://github.com/dotnet/roslyn-analyzers/blob/124afdcbb69d45e95794c8f289d3fc1f44c26518/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs#L96

When I check the type of the parameter, it's `INamedTypeSymbol` doesn't always match up with the well-known type, but the name of it always should.

Reply:
@stevenbrix I think that means you're somehow doing the comparison incorrectly. Probably you need `argumentOperation.Parameter.Type.OriginalDefinition.Equals(readOnlSpanSymbol)`. (The `OriginalDefinition` here gets the type before it was constructed, ie, if the type is for example `ReadOnlySpan<string>`, the original definition will be `ReadOnlySpan<T>`, which is what `GetOrCreateTypeByMetadataName` returns.

The problem with comparing type names is that the user can define a type with the same name. So symbol equality is always more reliable.

Reply:
Thanks. I'll work on those changes.